### DISCUSSION TITLE
Weird behavior on DEDENT token - the same token used twice

### DISCUSSION THREAD

**Discussion by m-kru** [tree-sitter/tree-sitter#1292] [Q&A] [answered]
tree-sitter version: 0.20.0 (e85a279cf29da1b08648e27214dda20a841e57c8)

I am prototyping some language where indent has semantic meaning and the indent is enforced to be the `\t` character. Dedent is of course the lack of `\t` character and one must keep track of the current indent.

The `dedent` token is used only within single rule:

```javascript
    element_body: $ => seq(
      $.newline,
      $.indent,
      repeat1(choice(
        $.element_definition,
        $._element_instantiation,
        $.property_assignment
      )),
      $.dedent
    ),
```

`element_body` is also used only within single rule:
``` javascript
    element_definition: $ => seq(
      $.element_type, $.identifier,
      choice(
        $.newline,
        seq(optional($.parameters_list), $.element_body)
      )
    ),
```
As one can see the structure can be nested as `element_body` can contain `element_definition` and vice versa. This implies that multiple dedent tokens can appear one after another.

For the following code:
```
func foo
	param bar
		width = 8
```

The S-expression looks like follows:
```
(description [0, 0] - [3, 0]
  (element_definition [0, 0] - [3, 0]
    (element_type [0, 0] - [0, 4])
    (identifier [0, 5] - [0, 8])
    (element_body [0, 8] - [3, 0]
      (newline [0, 8] - [1, 0])
      (indent [1, 0] - [1, 1])
      (element_definition [1, 1] - [3, 0]
        (element_type [1, 1] - [1, 6])
        (identifier [1, 7] - [1, 10])
        (element_body [1, 10] - [3, 0]
          (newline [1, 10] - [2, 0])
          (indent [2, 0] - [2, 2])
          (property_assignment [2, 2] - [3, 0]
            (identifier [2, 2] - [2, 7])
            (expression [2, 10] - [2, 11]
              (primary_expression [2, 10] - [2, 11]
                (decimal_literal [2, 10] - [2, 11])))
            (newline [2, 11] - [3, 0]))
          (dedent [3, 0] - [3, 0])))
      (dedent [3, 0] - [3, 0]))))
```
and my debug output looks like this:
```
1 I : lookahead 9 currnt indent 0
  I : current_indent 1
2 ID: lookahead 9 current indent 1
  I : current_indent 2
3  D: lookahead 0 current indent 2
   D: current indent 1
```
The same `dedent` token closes two `element_bodies`. This looks strange to me and I do not understand why tree-sitter does not scan for the second `dedent` token.

Now, for the following code:
```
func foo
	param bar
		width = 8
	<- here is one more '\t' character
```
the S-expression is as follows:
```
(description [0, 0] - [4, 0]
  (element_definition [0, 0] - [3, 1]
    (element_type [0, 0] - [0, 4])
    (identifier [0, 5] - [0, 8])
    (element_body [0, 8] - [3, 1]
      (newline [0, 8] - [1, 0])
      (indent [1, 0] - [1, 1])
      (element_definition [1, 1] - [3, 1]
        (element_type [1, 1] - [1, 6])
        (identifier [1, 7] - [1, 10])
        (element_body [1, 10] - [3, 1]
          (newline [1, 10] - [2, 0])
          (indent [2, 0] - [2, 2])
          (property_assignment [2, 2] - [3, 0]
            (identifier [2, 2] - [2, 7])
            (expression [2, 10] - [2, 11]
              (primary_expression [2, 10] - [2, 11]
                (decimal_literal [2, 10] - [2, 11])))
            (newline [2, 11] - [3, 0]))
          (dedent [3, 0] - [3, 1])))
      (dedent [3, 1] - [3, 1]))))
```
and my debug output looks like this:
```
1 I : lookahead 9 currnt indent 0
  I : current_indent 1
2 ID: lookahead 9 current indent 1
  I : current_indent 2
3  D: lookahead 9 current indent 2
   D: current indent 1
4  D: lookahead 10 current indent 1
   D: current indent 0
```
Now 2 distinct `dedent` tokens are used for `element_bodies`.

My external scanner scan functions is as follows:
```C
bool tree_sitter_fbdl_external_scanner_scan(
	void *payload,
	TSLexer *lexer,
	const bool *valid_symbols
) {
	static unsigned debug_cnt = 0;

	struct Scanner *scanner = payload;

	if (valid_symbols[INDENT] || valid_symbols[DEDENT]) {
		debug_cnt++;
		// Debug code
		if (valid_symbols[INDENT] && valid_symbols[DEDENT]) {
			printf("%d ID: lookahead %d current indent %d\n", debug_cnt, lexer->lookahead, scanner->current_indent);
		} else if (valid_symbols[DEDENT]) {
			printf("%d  D: lookahead %d current indent %d\n", debug_cnt, lexer->lookahead, scanner->current_indent);
		} else if (valid_symbols[INDENT]) {;
			printf("%d I : lookahead %d currnt indent %d\n", debug_cnt, lexer->lookahead, scanner->current_indent);
		}

		if (scanner->dedents > 0) {
			if (valid_symbols[DEDENT]) {
				goto dedent;
			}
		}

		unsigned indent = 0;
		while (lexer->lookahead == '\t') {
			indent++;
			lexer->advance(lexer, false);
		}

		if (valid_symbols[DEDENT] && (indent < scanner->current_indent)) {
			scanner->dedents = scanner->current_indent - indent;
dedent:
			scanner->current_indent--;
			scanner->dedents--;
			lexer->result_symbol = DEDENT;
			printf("   D: current indent %d\n", scanner->current_indent);
			return true;
		}

		if (valid_symbols[INDENT] && (indent > scanner->current_indent)) {
			if (indent != scanner->current_indent + 1) {
				errx(EXIT_FAILURE, "Multi indent detected, indent %d, current indent %d", indent, scanner->current_indent);
			}

			scanner->current_indent++;
			lexer->mark_end(lexer);
			lexer->result_symbol = INDENT;
			printf("  I : current_indent %d\n", scanner->current_indent);
			return true;
		}
	}

	if (valid_symbols[NEWLINE] && lexer->lookahead == '\n') {
		lexer->advance(lexer, false);

		while (lexer->lookahead == '\n') {
			lexer->advance(lexer, true);
		}

		lexer->mark_end(lexer);
		lexer->result_symbol = NEWLINE;
		return true;
	}

	return false;
}
```

As one can see the information about the multi dedent is stored within the scanner structure.
This strange behavior leads to a situation where following code is invalid:
```
func foo
	param bar
		width = 8

mask zaz
	atomic = false
```
but this one is valid:
```
func foo
	param bar
		width = 8
	<- extra '\t' character here

mask zaz
	atomic = false
```

**Comment by maxbrunsfeld** [comment]
I converted this to a discussion.

**Comment by m-kru** [comment]
@maxbrunsfeld @ahlinc @ahelwer I know you are quite fluent and active users of the tree-sitter. I would really appreciate if you could take a look at this issue. It has completely blocked me.

**Reply by ahelwer** [reply]
Haven't looked through the entire thing but I wrote up a fairly detailed post on my approach to parsing indentation blocks here: https://github.com/tree-sitter/tree-sitter/discussions/1215#discussioncomment-938407

tl;dr is I found it helpful to have `INDENT`/`NEWLINE`/`DEDENT` external tokens, where `NEWLINE` just separates elements of a single indented block. In general an indented block rule looks like:
```js
indented_block: $ => seq($.INDENT, repeat(seq($.statement, $.NEWLINE)), $.statement, $.DEDENT),
```
and I use a `std::vector<int16_t>` as a stack keeping track of column positions of the nested indent blocks. Since you want to limit leading line characters to `\t` instead of spaces you might have to write your own column position calculations instead of using the builtin `TSLexer` one.

**Reply by ahlinc** [reply]
@m-kru _I'm not sure that correctly understand your problem._

> Now 2 distinct dedent tokens are used for element_bodies.

I've looked to your S-exp like outputs and in both I see two `dedent` tokens:
_Excerpt from the first:_
```
(description [0, 0] - [3, 0]
  (element_definition [0, 0] - [3, 0]
    .....
    (element_body [0, 8] - [3, 0]
      (newline [0, 8] - [1, 0])
      (indent [1, 0] - [1, 1])                    <--- the outer indent
      (element_definition [1, 1] - [3, 0]
        .....
        (element_body [1, 10] - [3, 0]
          (newline [1, 10] - [2, 0])
          (indent [2, 0] - [2, 2])               <--- the inner indent
          (property_assignment [2, 2] - [3, 0]
            .....
            (newline [2, 11] - [3, 0]))
          (dedent [3, 0] - [3, 0])))             <--- the inner dedent
      (dedent [3, 0] - [3, 0]))))                <--- the outer dedent
```
_Excerpt from the second:_
```
(description [0, 0] - [4, 0]
  (element_definition [0, 0] - [3, 1]
    .....
    (element_body [0, 8] - [3, 1]
      (newline [0, 8] - [1, 0])
      (indent [1, 0] - [1, 1])                    <--- the outer indent
      (element_definition [1, 1] - [3, 1]
        .....
        (element_body [1, 10] - [3, 1]
          (newline [1, 10] - [2, 0])
          (indent [2, 0] - [2, 2])               <--- the inner indent
          (property_assignment [2, 2] - [3, 0]
            .....
            (newline [2, 11] - [3, 0]))
          (dedent [3, 0] - [3, 1])))             <--- the inner dedent
      (dedent [3, 1] - [3, 1]))))                <--- the outer dedent
```
The only difference that I see that dedents in the first case occupying the same range. And this looks ok cause there is no real tab symbol for the inner dedent in the source text. And in the second case the outer dedent has a zero length range cause there is no real tab symbol for it.

Could you explain how this affects your grammar or provide more detailed examples?

_Also it would be very helpful and time saving to have a link to a some repo with a minimal grammar + external scanner samples that demonstrate your issue._

### Edit 1
You can call the `id()` on your tree-sitter `Node` objects to verify that your `dedent` tokens are different instances.

**Comment by m-kru** [comment, marked as answer]
It looks like the source of the problem was the fact, that I have not been serializing and deserializing the scanner.

Nevertheless, I still do not fully understand what exactly was the problem.  The logs I have added to the scanner (simple `printf`s) did not show that scanner was changing state, and that then it was not restored correctly. However, after adding serialization/deserialization everything works as expected.
